### PR TITLE
add Uni Match v2 trainer

### PR DESCRIPTION
**Pull request by Marei33**
WIP

### CODE REVIEW

**Comment by Marei33** [comment]
wait for PR #664 to be merged first

**Comment by Marei33** [comment]
@constantinpape @anwai98 then this PR is also ready to review!

File: torch_em/self_training/logger.py [hunk 1]
```diff
@@ -169,4 +169,230 @@ def log_validation_augmentations(
                 xu2,
                 pseudo_labels,
                 pred,
-            )
\ No newline at end of file
+            )
+
+
+class UniMatchv2TensorboardLogger(torch_em.trainer.logger_base.TorchEmLogger):
+    """Logger for self-training via `torch_em.self_training.FixMatch` or `torch_em.self_training.MeanTeacher`.
```
Comment (constantinpape) [inline torch_em/self_training/logger.py]: The doc doesn't make sense, I guess this is specifically for the UniMatch trainer?!

File: torch_em/self_training/loss.py [hunk 2]
```diff
@@ -251,4 +251,104 @@ def __call__(
...
+        label_filter: Optional[torch.Tensor] = None,
+        pred_dim: int = 1,
+    ) -> torch.Tensor:
+        """Compute the loss for self-training.
+
+        Args:
+            model: The model.
+            input_: The model inputs for this batch.
+            labels: The (pseudo) labels for this batch.
+            label_filter: A mask to exclude from the loss computation.
```
Comment (constantinpape) [inline torch_em/self_training/loss.py]: Please also document `pred_dim` here and explain why it's needed / when to use it.

File: torch_em/self_training/loss.py [hunk 3]
```diff
@@ -251,4 +251,104 @@ def __call__(
...
+
+        Returns:
+            The loss value.
+        """
+        if self.activation is not None:
+            prediction = self.activation(prediction)
+
+        if pred_dim == 2:
```
Comment (constantinpape) [inline torch_em/self_training/loss.py:318]: Please add a comment with an explanation to why this is needed / what the rationale is. This is not very obvious.

File: torch_em/self_training/loss.py [hunk 4]
```diff
@@ -251,4 +251,104 @@ def __call__(
...
+            if label_filter is None:
+                loss = (self.loss(prediction[0], labels) + self.loss(prediction[1], labels)) / 2
+            else:
+                loss = (self.loss(
+                    prediction[0] * label_filter, labels * label_filter
+                ) + self.loss(prediction[1] * label_filter, labels * label_filter)) / 2
+            return loss
+
+        else:
```
Comment (constantinpape) [inline torch_em/self_training/loss.py:327]: It seems like only `pred_dim == 1` or `pred_dim == 2` make sense. But currently your code runs for any value of pred_dim. I suggest to update this to raise a value error for invalid `pred_dim` values.

File: torch_em/self_training/loss.py [hunk 5]
```diff
@@ -251,4 +251,104 @@ def __call__(
...
+        metric: nn.Module = DiceLoss(),
+        activation: Optional[nn.Module] = None
+    ):
+        super().__init__()
+        self.activation = activation
+        self.loss = loss
+        self.metric = metric
+        self.init_kwargs = {}
+
+    def __call__(
```
Comment (constantinpape) [inline torch_em/self_training/loss.py:361]: See comments on `pred_dim` above.

File: torch_em/self_training/loss.py (hunk 4 above)
Comment (Marei33) [reply torch_em/self_training/loss.py:327]: Yes, makes sense, added that to the loss function

**Comment by Marei33** [comment]
@constantinpape updated all the mentioned docstrings + the UniMatchv2Trainer docstring

**Review by constantinpape — APPROVED** [review]
Looks good!
